### PR TITLE
[YUNIKORN-2172] Add test cases for different applications each with different group linkage

### PR DESCRIPTION
**Pull request by chenyulin0719**
### What is this PR for?
Add test cases to cover the todo action. 
`//nolint: todo test cases for different applications each with different group linkage
`
Increase resource to ugm for different applications each have different group linkage.

New tests:
- TestIncreaseDifferentAppResourceToDiffGroup() in  manager_test.go
- TestSetGroupForApp() in user_tracker_test.go

### What type of PR is it?
* [ ] - Bug Fix
* [x] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
NA

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-2172

### How should this be tested?
All make test should pass.

### Screenshots (if appropriate)
NA
### Questions:
NA

### CODE REVIEW

File: pkg/scheduler/ugm/manager_test.go [hunk 1]
```diff
@@ -702,6 +702,62 @@ func TestDecreaseTrackedResourceForGroupTracker(t *testing.T) {
...
+	assert.Equal(t, len(userTracker.appGroupTrackers), 0, "Inital state shouldn't hold any group tracker object in UserTracker.")
+	assert.Equal(t, len(groupTracker1.applications), 0, "Initial state shouldn't have any application in GroupTracker.")
+	assert.Equal(t, len(groupTracker2.applications), 0, "Initial state shouldn't have any application in GroupTracker.")
+	assert.Assert(t, userTracker.queueTracker.resourceUsage.IsEmpty(), "Initial state shouldn't have any resources tracked in UserTracker.")
+	assert.Assert(t, groupTracker1.queueTracker.resourceUsage.IsEmpty(), "Initial state shouldn't have any resources tracked in GroupTracker.")
+	assert.Assert(t, groupTracker2.queueTracker.resourceUsage.IsEmpty(), "Initial state shouldn't have any resources tracked in GroupTracker.")
+
+	// Increase resources to different application, different group, same user
+	userGroup1 := security.UserGroup{User: user, Groups: []string{"group1"}}
+	userGroup2 := security.UserGroup{User: user, Groups: []string{"group2"}}
```
[the remark this answers is not in the document]
Comment (chenyulin0719) [reply pkg/scheduler/ugm/manager_test.go]: Hi @manirajv06,
Thanks for providing the real world test case. Just update test as suggested. 06bab12b29c714231e00e6f14816c5cf4bc40254

If you could review the the change, I would greatly appreciate it.  Thanks. :)

File: pkg/scheduler/ugm/manager_test.go [hunk 2]
```diff
@@ -702,6 +702,64 @@ func TestDecreaseTrackedResourceForGroupTracker(t *testing.T) {
 	assert.Equal(t, resources.Equals(groupTracker.queueTracker.childQueueTrackers["parent"].resourceUsage, resources.Zero), true)
 }
 
+func TestIncreaseDifferentAppResourceToDiffGroupLinkage(t *testing.T) {
```
Comment (manirajv06) [inline pkg/scheduler/ugm/manager_test.go:705]: Can we rename it to TestUserGroupLimitWithMultipleApps ?

File: pkg/scheduler/ugm/manager_test.go [hunk 3]
```diff
@@ -702,6 +702,64 @@ func TestDecreaseTrackedResourceForGroupTracker(t *testing.T) {
...
+	assert.Assert(t, groupTracker2.queueTracker.resourceUsage.IsEmpty(), "Initial state shouldn't have any resources tracked in GroupTracker.")
+
+	// Increase resources to different application, different queue
+	userGroup := security.UserGroup{User: user, Groups: groups}
+	usage, err := resources.NewResourceFromConf(map[string]string{"memory": "10", "vcore": "10"})
+	if err != nil {
+		t.Errorf("new resource create returned error or wrong resource: error %t, res %v", err, usage)
+	}
+	m.IncreaseTrackedResource("root.parent.leaf1", TestApp1, usage, userGroup)
+	m.IncreaseTrackedResource("root.parent.leaf2", TestApp2, usage, userGroup)
```
Comment (manirajv06) [inline pkg/scheduler/ugm/manager_test.go:741]: Try doing further increase and ensure limit has been limit. Then, try to remove app through decrease case and ensure entry has been removed in corresponding maps.

File: pkg/scheduler/ugm/manager_test.go [hunk 4]
```diff
@@ -702,6 +702,64 @@ func TestDecreaseTrackedResourceForGroupTracker(t *testing.T) {
...
+	assert.Equal(t, userTracker.appGroupTrackers[TestApp1], groupTracker2, "UserTracker should hold application/GroupTracker mapping.")
+	assert.Equal(t, userTracker.appGroupTrackers[TestApp2], groupTracker1, "UserTracker should hold application/GroupTracker mapping.")
+
+	// Assert GroupTracker post state
+	assert.Assert(t, resources.Equals(groupTracker1.queueTracker.resourceUsage, usage), "Should have new resource tracked in GroupTracker.")
+	assert.Assert(t, resources.Equals(groupTracker2.queueTracker.resourceUsage, usage), "Should have new resource tracked in GroupTracker.")
+	assert.Equal(t, groupTracker1.applications[TestApp2], userGroup.User, "GroupTracker should hold applications/user mapping.")
+	assert.Equal(t, groupTracker2.applications[TestApp1], userGroup.User, "GroupTracker should hold applications/user mapping.")
+	assert.Assert(t, resources.Equals(groupTracker1.queueTracker.resourceUsage, usage), "The new increased resource should be tracked by GroupTracker.")
+	assert.Assert(t, resources.Equals(groupTracker2.queueTracker.resourceUsage, usage), "The new increased resource should be tracked by GroupTracker.")
```
Comment (manirajv06) [inline pkg/scheduler/ugm/manager_test.go:760]: I can see a lot of asserts. Try to keep only mandatory checks or create a common method to do the assertion by passing necessary values as arguments.

Comment (manirajv06) [reply pkg/scheduler/ugm/manager_test.go:760]: Most important thing to assert is, resource usage for either root.parent or root with different groups like group 1 and group2. In addition, there should not be group tracking for root.parent.leaf2, but group2 should be tracked for root.parent.leaf1

**Review by manirajv06 — CHANGES_REQUESTED** [review]

File: pkg/scheduler/ugm/manager_test.go (hunk 3 above)
Comment (chenyulin0719) [reply pkg/scheduler/ugm/manager_test.go]: Update in 97d66c8beb6a87d02e6ff3d233012b2f8cafb45b ([Line 766~789](https://github.com/apache/yunikorn-core/commit/97d66c8beb6a87d02e6ff3d233012b2f8cafb45b#diff-31ea862574435965fec182f9616821c52ff5c57a5df0c2e984d9b8ab52a41d21R766-R789))

File: pkg/scheduler/ugm/manager_test.go [hunk 5]
```diff
@@ -702,6 +702,93 @@ func TestDecreaseTrackedResourceForGroupTracker(t *testing.T) {
...
+	// Assert ugm/GroupTracker state after apps removed, UserTracker should be removed.
+	assertUgmTrackedResourceCount(t, 0, 2)
+	assertGroupTrackerResources(t, "group1", TestApp2, emptyUsage, 0, "")
+	assertGroupTrackerResources(t, "group2", TestApp1, emptyUsage, 0, "")
+
+	// root.parent.leaf1 TestApp1 -> was limited by group 2, root.parent.leaf1
+	// root.parent.leaf2 TestApp2 -> was limited by group 1, root.parent
+	assertGroupTrackerQueuePathResource(t, "group1", "root.parent.leaf1", nil)
+	// assertGroupTrackerQueuePathResource(t, "group1", "root.parent.leaf2", emptyUsage)  // -> Failed
+	// assertGroupTrackerQueuePathResource(t, "group1", "root.parent.leaf2", usage)       // -> Passed
```
Comment (chenyulin0719) [inline pkg/scheduler/ugm/manager_test.go:783]: Hi @manirajv06,

I encountered a failed assert in Line 782 after decreasing resources, 
and I believe this is a bug related GroupTracker.DecreaseTrackedResource().

When I increased the resource in the 'root.parent.leaf2' queue,  (the limit was set on 'root.parent')
the resource was tracked in the following queueTracker:
- root.parent.leaf2
- root.parent
- root

However, when I decreased the resource from 'root.parent.leaf2',
it only removed resource in the following queueTracker:
- root.parent
- root

The root cause might be that there is no limit on 'root.parent.leaf2'.
The traversal of decreasing only decrease resources before root.parenet

Please correct me if I'm wrong.

Comment (chenyulin0719) [reply pkg/scheduler/ugm/manager_test.go:783]: Sorry, please let me correct myself.  

- The behavior of GroupTracker.DecreaseTrackedResource() is correct
- The wrong state happened on QueueTracker.increaseTrackedResource() if the parent queue reach the limitation. for example: 

Before state:
- root.parent           (usage 10 memory, 10 vcore) + (limit 10 memory, 10 vcore)
- root.parent.leaf2 (usage 10 memory, 10 vcore) + (no limit)

After adding (10 memory, 10 vcore) resource to  root.parent.leaf2
- root.parent           (usage 10 memory, 10 vcore) 
- root.parent.leaf2.(usage 20 memory, 20 vcore)

Comment (chenyulin0719) [reply pkg/scheduler/ugm/manager_test.go:783]: Maybe it's intentional because ugh doesn't care about those leaf queues without limits?

Comment (chenyulin0719) [reply pkg/scheduler/ugm/manager_test.go:783]: Summary the issue:
Queue Setup:
- **root.parent with group1 limit**
- **root.parent.leaf2 with no limit**

Before assert:
- increaseTrackedResourceThenAssert(t, "root.parent.leaf2", TestApp2, usage, userGroup, true) //**success**
- increaseTrackedResourceThenAssert(t, "root.parent.leaf2", TestApp2, usage, userGroup, false)  
 **// reach 'root.parent' limit, but it still increase root.parent.leaf2's usage.**
- decreaseTrackedResourceThenAssert(t, "root.parent.leaf2", TestApp2, usage, userGroup, true, true) //**success**

Assert:
- assertGroupTrackerQueuePathResource(t, "group1", "root.parent.leaf2", emptyUsage) 
  **// -> Failed.  (Should Pass)**
- assertGroupTrackerQueuePathResource(t, "group1", "root.parent.leaf2", usage)              
  **//-> Passed. (Should Failed)**

**Comment by manirajv06** [comment]
@chenyulin0719 Overall, I would say change is huge. I don't think we need this much. Try to keep it simple and neat. Please refer the other tests and see how tests have been developed. I would do something like below. Run different app with different groups, assert resource usage, remove the apps and repeat asserts. thats it.

func TestUserGroupLimitWithMultipleApps(t *testing.T) {
	// Setup
	setupUGM()
	manager := GetUserManager()
	userGroup := security.UserGroup{User: "user", Groups: []string{"group1", "group2"}}

	conf := createConfigWithDifferentGroupsLinkage(userGroup.Groups[0], userGroup.Groups[1], 10, 10, 3)
	assert.NilError(t, m.UpdateConfig(conf.Queues[0], "root"))

	usage, err := resources.NewResourceFromConf(map[string]string{"memory": "10", "vcores": "10"})
	if err != nil {
		t.Errorf("new resource create returned error or wrong resource: error %t, res %v", err, usage)
	}

        // run different apps with different groups
	increased := manager.IncreaseTrackedResource(queuePath1, TestApp1, usage, userGroup)
	assert.Equal(t, increased, true)
	increased = manager.IncreaseTrackedResource(queuePath2, TestApp2, usage, userGroup)
	assert.Equal(t, increased, true)

	// ensure different groups are linked and resource usage is correct
	assert.Equal(t, len(manager.getUserTracker("user").appGroupTrackers), 2)
	gt1 := manager.getUserTracker("user").appGroupTrackers[TestApp1]
	gt2 := manager.getUserTracker("user").appGroupTrackers[TestApp2]
	assert.Equal(t, gt1.groupName, "group2")
	assert.Equal(t, resources.Equals(gt1.queueTracker.resourceUsage, usage), true)
	assert.Equal(t, gt2.groupName, "group1")
	assert.Equal(t, resources.Equals(gt2.queueTracker.resourceUsage, usage), true)

	// limit has reached for both the groups
	increased = manager.IncreaseTrackedResource(queuePath1, TestApp1, usage, userGroup)
	assert.Equal(t, increased, false)
	increased = manager.IncreaseTrackedResource(queuePath2, TestApp2, usage, userGroup)
	assert.Equal(t, increased, false)

        // remove the apps
	decreased := manager.DecreaseTrackedResource(queuePath1, TestApp1, usage, userGroup, true)
	assert.Equal(t, decreased, true)
	decreased = manager.DecreaseTrackedResource(queuePath2, TestApp2, usage, userGroup, true)
	assert.Equal(t, decreased, true)
      
        // assert group linkage has been removed
	assert.Equal(t, len(manager.getUserTracker("user").appGroupTrackers), 0)
}

func createConfigWithDifferentGroupsLinkage(group1 string, group2 string, mem int, vcore int, maxApps uint64) configs.PartitionConfig {
	// root.parent with group1 limit
	// root.parent.leaf1 with group2 limit
	// root.parent.leaf2 with no limit
	conf := configs.PartitionConfig{
		Name: "test",
		Queues: []configs.QueueConfig{
			{
				Name:      "root",
				Parent:    true,
				SubmitACL: "*",
				Queues: []configs.QueueConfig{
					{
						Name:      "parent",
						Parent:    true,
						SubmitACL: "*",
						Queues: []configs.QueueConfig{
							{
								Name:      "child1",
								Parent:    false,
								SubmitACL: "*",
								Queues:    nil,
								Limits: []configs.Limit{
									{
										Limit: "leaf queue with group2 limit",
										Groups: []string{
											group2,
										},
										MaxResources: map[string]string{
											"memory": strconv.Itoa(mem),
											"vcores": strconv.Itoa(vcore),
										},
										MaxApplications: maxApps,
									},
								},
							},
							{
								Name:      "child2",
								Parent:    false,
								SubmitACL: "*",
								Queues:    nil,
							},
						},
						Limits: []configs.Limit{
							{
								Limit: "parent queue with group1 limit",
								Groups: []string{
									group1,
								},
								MaxResources: map[string]string{
									"memory": strconv.Itoa(mem),
									"vcores": strconv.Itoa(vcore),
								},
								MaxApplications: maxApps,
							},
						},
					},
				},
			},
		},
	}
	return conf
}

File: pkg/scheduler/ugm/manager_test.go (hunk 3 above)
Comment (chenyulin0719) [reply pkg/scheduler/ugm/manager_test.go]: Update to the simple test case.

File: pkg/scheduler/ugm/manager_test.go (hunk 4 above)
Comment (chenyulin0719) [reply pkg/scheduler/ugm/manager_test.go]: Update to the simple test case as suggested below.  Remove the t.help functions I created in the last PR.

Comment (chenyulin0719) [reply pkg/scheduler/ugm/manager_test.go]: > Most important thing to assert is, resource usage for either root.parent or root with different groups like group 1 and group2. In addition, there should not be group tracking for root.parent.leaf2, but group2 should be tracked for root.parent.leaf1

The current version: assert the 'root' queueTracker resourceUsage for different groups.
I remove the assertGroupTrackerQueuePathResource() function I created in the last PR to make things simpler.

**Comment by chenyulin0719** [comment]
Hi @manirajv06 , thanks for providing the example.
Just updated the test case as suggested.

File: pkg/scheduler/ugm/user_tracker_test.go [hunk 6]
```diff
@@ -224,6 +222,17 @@ func TestSetMaxLimits(t *testing.T) {
...
+func TestSetGroupForApp(t *testing.T) {
+	userTracker := newUserTracker("user")
+	groupTracker1 := newGroupTracker("group1")
+	groupTracker2 := newGroupTracker("group2")
+	userTracker.setGroupForApp(TestApp1, groupTracker1)
+	userTracker.setGroupForApp(TestApp2, groupTracker2)
+
+	assert.Equal(t, 2, len(userTracker.getTrackedApplications()))
+	assert.Equal(t, 2, len(userTracker.appGroupTrackers))
+}
```
Comment (manirajv06) [inline pkg/scheduler/ugm/user_tracker_test.go:234]: Do we really need this new test? I think we can move these two asserts after line no. 76 in earlier method?

File: pkg/scheduler/ugm/manager_test.go [hunk 7]
```diff
@@ -1312,11 +1367,11 @@ func createUpdateConfigWithWildCardUsersAndGroups(user string, group string, wil
...
 }
 
-func createUpdateConfig(user string, group string) configs.PartitionConfig {
-	return createConfig(user, group, "memory", "10", 50, 5)
+func createUpdateConfig(user string, groups []string) configs.PartitionConfig {
+	return createConfig(user, groups, "memory", "10", 50, 5)
 }
 
-func createConfig(user string, group string, resourceKey string, resourceValue string, mem int, maxApps uint64) configs.PartitionConfig {
+func createConfig(user string, groups []string, resourceKey string, resourceValue string, mem int, maxApps uint64) configs.PartitionConfig {
```
Comment (manirajv06) [inline pkg/scheduler/ugm/manager_test.go:1374]: Do we really need this change as part of this pr? 

If no, can we take it up in different jira? we can change user too.

**Review by manirajv06 — CHANGES_REQUESTED** [review]

File: pkg/scheduler/ugm/manager_test.go (hunk 7 above)
Comment (chenyulin0719) [reply pkg/scheduler/ugm/manager_test.go]: To keep this PR pure simple,  removed the refactor code.

File: pkg/scheduler/ugm/user_tracker_test.go (hunk 6 above)
Comment (chenyulin0719) [reply pkg/scheduler/ugm/user_tracker_test.go]: My original thought was to add unit test that can partially address the previous todo comment. 
//nolint: todo test cases for different applications each with different group linkage

Just removed it to simplify this PR.

**Comment by chenyulin0719** [comment]
Hi @manirajv06,
Just removed the unneccesarily changes as suggested.

**Review by manirajv06 — APPROVED** [review]

**Review by manirajv06 — APPROVED** [review]

**Pull request closed without merging**